### PR TITLE
feat:delete environment command

### PR DESCRIPTION
**Pull request by shivam-Purohit**
Currently there is no command to delete the chaos environment. This feature will aim to add command to delete chaos environments based on the id provided.

### CODE REVIEW

File: pkg/cmd/delete/environment.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+		utils.White_B.Println("\n🚀 Chaos Environment successfully deleted.")
+
+	},
+}
+
+func init() {
+	DeleteCmd.AddCommand(environmentCmd)
+
+	environmentCmd.Flags().String("project-id", "", "Set the project-id to delete Chaos Envuronment for the particular project. To see the projects, apply litmusctl get projects")
```
Comment (neelanjan00) [inline pkg/cmd/delete/environment.go:136]:
```suggestion
	environmentCmd.Flags().String("project-id", "", "Set the project-id to delete Chaos Environment for the particular project. To see the projects, apply litmusctl get projects")
```

**Comment by SarthakJain26** [comment]
@Shivam you will have to add a check to see if there are any infras that are present for the environment, if yes you can just prompt that there are infras present in the environment and they need to be deleted before deleting the environment.

**Comment by shivam-Purohit** [comment]
@SarthakJain26 @neelanjan00 
Currently i am an  issue. If i delete a environment and try to create a new environment immediately then this error occurs
`Chaos Environment connection failed: write exception: write errors: [E11000 duplicate key error collection: litmus.environment index: environment_id_1 dup key: { environment_id: "env1" }]`
does this is not reflecting in the db?


also we are using `LIST `environment. But as i checked there already exist a handlers for get environment in graphlql. I suggest we should use that for `get chaos-environment` command rather than iterating through list of environments.This way we would provide more details and isolate both commands, this will also help in testing commands separately. We should make changes in the `get chaos-environment` command and` list chaos-environments `command

**Comment by SarthakJain26** [comment]
> @SarthakJain26 @neelanjan00 Currently i am an issue. If i delete a environment and try to create a new environment immediately then this error occurs `Chaos Environment connection failed: write exception: write errors: [E11000 duplicate key error collection: litmus.environment index: environment_id_1 dup key: { environment_id: "env1" }]` does this is not reflecting in the db?
> 
> also we are using `LIST `environment. But as i checked there already exist a handlers for get environment in graphlql. I suggest we should use that for `get chaos-environment` command rather than iterating through list of environments.This way we would provide more details and isolate both commands, this will also help in testing commands separately. We should make changes in the `get chaos-environment` command and`list chaos-environments`command
> 
> ```
> 	var isEnvExist = false
> 	for i := range Env.Data.ListEnvironmentDetails.Environments {
> 		if newInfra.EnvironmentID == Env.Data.ListEnvironmentDetails.Environments[i].EnvironmentID {
> 			utils.White_B.Println(Env.Data.ListEnvironmentDetails.Environments[i].EnvironmentID)
> 			isEnvExist = true
> 			break
> 		}
> 	}
> ```
> 
> This is other instance we are using the iteration which imo should not be needed.

@shivam-Purohit On deleting it does get reflected in the DB but the error you are seeing is because we are not deleting the document from the DB rather we are just marking it as removed, and since an mongo index is set on name field, therefore mongo does not allows the creation of a new environment with the same name. The solution to this is make the index as partial index and not consider the documents that are marked as removed. Anyways this is a separate issue which will have to be updated in graphql code.

Regarding using GET command instead of LIST, you are right, we should use GET where we want to fetch just one resource and we should be using LIST where multiple resources are being fetched. You can go ahead and update the implementation. Thanks

**Comment by shivam-Purohit** [comment]
@SarthakJain26 so what should be the next step in this feature then?

Also I have pushed the changes I suggested if you can take a look.

File: pkg/cmd/delete/environment.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+		}
+		if !editAccess {
+			utils.Red.Println("⛔ User doesn't have edit access to the project!!")
+			os.Exit(1)
+		}
+
+		environmentList, err := environment.GetEnvironmentList(projectID, credentials)
+		if err != nil {
+			if strings.Contains(err.Error(), "permission_denied") {
+				utils.Red.Println("❌ You don't have enough permissions to access this resource.")
```
Comment (SarthakJain26) [inline pkg/cmd/delete/environment.go:108]:
```suggestion
				utils.Red.Println("❌ You don't have enough permissions to delete an environment.")
```

File: pkg/cmd/delete/environment.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+				utils.PrintError(err)
+				os.Exit(1)
+			}
+		}
+		environmentListData := environmentList.Data.ListEnvironmentDetails.Environments
+		for i := 0; i < len(environmentListData); i++ {
+			if environmentListData[i].EnvironmentID == environmentID {
+				if len(environmentListData[i].InfraIDs) > 0 {
+					utils.Red.Println("Chaos Infras present in the Chaos Environment")
+					utils.Red.Println("Delete the Chaos Infras first to delete the Environment")
```
Comment (SarthakJain26) [inline pkg/cmd/delete/environment.go:120]: Can you try to club these messages in a single statement

File: pkg/cmd/delete/environment.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+				os.Exit(1)
+			} else {
+				utils.PrintError(err)
+				os.Exit(1)
+			}
+		}
+		environmentListData := environmentList.Data.ListEnvironmentDetails.Environments
+		for i := 0; i < len(environmentListData); i++ {
+			if environmentListData[i].EnvironmentID == environmentID {
+				if len(environmentListData[i].InfraIDs) > 0 {
```
Comment (SarthakJain26) [inline pkg/cmd/delete/environment.go:118]: @Saranya-jena can you once confirm if we are deleting the infras from the infra array in environments collection on infra deletion

**Review by SarthakJain26 — CHANGES_REQUESTED** [review]
@shivam-Purohit please also add the example usage of command in delete/delete.go and Usage_0.23.0.md file

File: pkg/cmd/delete/environment.go (hunk 4 above)
Comment (Saranya-jena) [reply pkg/cmd/delete/environment.go:118]: yes we are!

**Comment by shivam-Purohit** [comment]
> @shivam-Purohit please also add the example usage of command in delete/delete.go and Usage_0.23.0.md file

done! can you take another look? thank you!

**Comment by shivam-Purohit** [comment]
> The solution to this is make the index as partial index and not consider the documents that are marked as removed.

I see a commit pushed in litmus regarding this. Should this issue now be fixed @SarthakJain26 ?

**Comment by SarthakJain26** [comment]
> > The solution to this is make the index as partial index and not consider the documents that are marked as removed.
> 
> I see a commit pushed in litmus regarding this. Should this issue now be fixed @SarthakJain26 ?

Yes Shivam! This will be fixed now.

**Review by SarthakJain26 — APPROVED** [review]

**Review by Saranya-jena — APPROVED** [review]

File: Usage_0.23.0.md [hunk 5]
```diff
@@ -313,6 +313,20 @@ Enter the Environment Name: test2
...
+```shell
+litmusctl delete chaos-environment
+
+Enter the Project ID: eb7fc0a0-5878-4454-a9db-b67d283713bc
+
+Enter the Environment ID: testenv
+
+Are you sure you want to delete this Chaos Environment? (y/n):y
+
+🚀 Chaos Environment successfully deleted.
```
Comment (Jonsy13) [inline Usage_0.23.0.md:327]: To keep it generic with other logs in this file (like creation)
```suggestion
🚀 Chaos Environment deleted successfully.
```

**Comment by shivam-Purohit** [comment]
@SarthakJain26 done!

**Pull request merged**